File: 17. Letter Combinations of a Phone Number/Recursive Writing/letter_combinations_of_a_phone_number.py
combinations= [" ", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"]

def word_from_odo (relevant_digits, digits_odo):
	out = ""
	
	for i in range(len(relevant_digits)):
		out += combinations[int(relevant_digits[i])][digits_odo[i]]
	
	return out

def word_fill (relevant_digits, digits_odo, depth, max_depth, sample_space):
	#max_depth is the length of the relevant_digits
	letter_options = combinations[int(relevant_digits[depth])]
	len_letter_options = len(letter_options)
	
	for i in range(len_letter_options):
		digits_odo[depth] = i
		if (depth < (max_depth - 1)):
			word_fill(relevant_digits, digits_odo, depth + 1, max_depth, sample_space)
		else:
			##  sampe_space[sample_space_cursor] = ""
			sample_space.append(word_from_odo(relevant_digits, digits_odo))
			#should be append for the dynamic array programming languages


def get_letter_combinations (digits):
	out = []
	
	len_digits = len(digits)
	odo = []
	
	useable_digits = ""
	sample_size = 0
	for i in range(len_digits):
		no_of_letters = len(combinations[int(digits[i])])
		
		if no_of_letters == 0:
			continue
		
		useable_digits += digits[i]
		
		if (sample_size == 0):
			sample_size = len(combinations[int(digits[i])])
		else:
			sample_size *= len(combinations[int(digits[i])])
		
		odo.append(0)
	
	print("The useable digits are\t" + str(useable_digits))
	print("The size of the combinations is\t" + str(sample_size))
	len_useable_digits = len(useable_digits)
	word_fill (useable_digits, odo, 0, len_useable_digits, out)
	
	return out
	
digits = "6872"
print(get_letter_combinations(digits))

File: 17. Letter Combinations of a Phone Number/Recursive Writing/letter_combinations_of_a_phone_number.js
let combinations= [" ", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"]

function wordFromOdo (relevantDigits, digitsOdo){
	let out = "";
	
	for (let i = 0; i < relevantDigits.length; i++){
		out += combinations[Number(relevantDigits[i])][digitsOdo[i]];
	}
	
	return out;
}

function wordFill (relevantDigits, digitsOdo, depth, maxDepth, sampleSpace){
	let letterOptions = combinations[Number(relevantDigits[depth])];
	let lenLetterOptions = letterOptions.length;
	
	for (let i = 0; i < lenLetterOptions; i++){
		digitsOdo[depth] = i;
		if (depth < (maxDepth - 1)){
			wordFill (relevantDigits, digitsOdo, (depth + 1), maxDepth, sampleSpace);
		} else {
			sampleSpace.push(wordFromOdo(relevantDigits, digitsOdo));
		}
	}
}

function getLetterCombinations (digits){
	let out = [];
	
	let lenDigits = digits.length;
	let odo = [];
	
	let useableDigits = "";
	let sampleSize = 0;
	
	for (let i = 0; i < lenDigits; i++){
		if (combinations[Number(digits[i])] == 0){
			continue;
		}
		
		useableDigits += digits[i];
		odo.push(0);
	}
	
	let lenUseableDigits = useableDigits.length;
	
	wordFill (useableDigits, odo, 0, lenUseableDigits, out);
	
	return (out);
}

digits = "6872"
console.log(getLetterCombinations(digits))
